nkms_eth/base.py

```diff
@@ -0,0 +1,137 @@
+from abc import ABC, abstractmethod
+
+from nkms_eth.blockchain import TheBlockchain
+
+
+class Actor(ABC):
+    def __init__(self, address):
+        if isinstance(address, bytes):
+            address = address.hex()
+        self.address = address
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        r = "{}(address='{}')"
+        r.format(class_name, self.address)
+        return r
+
+
+class ContractDeployer(ABC):
+
+    _contract_name = NotImplemented
+
+    class ContractDeploymentError(Exception):
+        pass
+
+    def __init__(self, blockchain):
+        self.__armed = False
+        self._contract = None
+
+        # Sanity check
+        if not isinstance(blockchain, TheBlockchain):
+            error = 'Only TheBlockchain can be used to create a deployer, got {}.'
+            raise ValueError(error.format(type(blockchain)))
+        self._blockchain = blockchain
+
+    def __eq__(self, other):
+        return self._contract.address == other.address
+
+    @property
+    def contract_address(self) -> str:
+        try:
+            address = self._contract.address
+        except AttributeError:
+            cls = self.__class__
+            raise cls.ContractDeploymentError('Contract not deployed')
+        else:
+            return address
+
+    @property
+    def is_deployed(self) -> bool:
+        return bool(self._contract is not None)
+
+    @property
+    def is_armed(self) -> bool:
+        return bool(self.__armed is True)
```

Comment:
What's armed and how is it used (if)?

Reply:
Yeah - Need more docstrings here. The idea is that a `ContractDeployer` must be armed (like a safety switch) before it can deploy an ethereum smart contract. Calling `ContractDeployer.deploy` will raise a `ContractDeploymentError` if it has not been armed first by calling `ContractDeployer.arm`.  This adds some flexibility, and safety for handling contract wrapper types

Reply:
True. Admittedly - I have done this in tests... Ideally we can interact with a deployer in a few other ways other than deploying to the blockchain (i.e. mutate or call methods on `ContractDeployer` subclasses).  When several contract wrapper types that depend on one another need to be scripted together, it's nice and semantic to split up setup and deploy logic.  Also, a `ContractAgent` holds reference to a single `ContractDeployer`, for access to deployer constants.

Reply:
Also - `ContractDeployer` returns `None`, preventing method chaining.